core/src/main/java/io/questdb/cairo/wal/WalEventReader.java

```diff
@@ -77,21 +81,29 @@ public WalEventCursor of(Path path, long segmentTxn) {
 
             if (segmentTxn > -1) {
                 // Read record offset and size
-                long fdi = openRO(ff, path.trimTo(pathLen).concat(EVENT_INDEX_FILE_NAME).$(), LOG);
+                eventIndexMem.of(
+                        ff,
+                        path.trimTo(pathLen).concat(EVENT_INDEX_FILE_NAME).$(),
+                        ff.getPageSize(),
+                        -1,
+                        MemoryTag.MMAP_TABLE_WAL_READER,
+                        CairoConfiguration.O_NONE,
+                        Files.POSIX_MADV_RANDOM
+                );
                 try {
-                    int maxTxn = eventMem.getInt(WALE_MAX_TXN_OFFSET_32);
-                    long offset = ff.readNonNegativeLong(fdi, segmentTxn << 3);
-                    long size = ff.readNonNegativeLong(fdi, (maxTxn + 1L) << 3);
+                    final int maxTxn = eventMem.getInt(WALE_MAX_TXN_OFFSET_32);
+                    final long offset = readNonNegativeLong(eventIndexMem, segmentTxn << 3);
+                    long size = readNonNegativeLong(eventIndexMem, (maxTxn + 1L) << 3);
 
                     if (offset > -1 && size < WALE_HEADER_SIZE + Integer.BYTES) {
                         // index file may not contain all records from data file, but it should contain
                         // the transaction we need to read, e.g. segmentTxn
-                        size = ff.readNonNegativeLong(fdi, (segmentTxn + 1L) << 3);
+                        size = readNonNegativeLong(eventIndexMem, (segmentTxn + 1L) << 3);
```

Comment:
Earlier `_events.i` is mapped to the size that is returned from `Files.length()` because of `-1` passed as a parameter, and it can be smaller than what we want to read here. Please check if `eventIndexMem.size() < (segmentTxn + 2L) << 3` and remap it to that length.
`offset` should also be re-read if the mapping was smaller than expected.

Reply:
Logic rewritten and re-commented.